### ISSUE TITLE
Custom Inspectors

### ISSUE THREAD

**Issue by ChrisA89** [sksamuel/scalac-scapegoat-plugin#97]
I know you're keen for people to add useful Inspectors back to the project, but would it be possible to allow creation of custom Inspectors outside of the scapegoat codebase?

**Comment by sksamuel** [comment]
Yes, that's a good idea. Any particular preference on how it would work.

**Comment by ChrisA89** [comment]
I've added a pull request. I haven't tested that it works, but it's the sort of way that I imagine adding extra Inspections.

(The sbt plugin could enforce type safety where it declares the list).

**Comment by RichardBradley** [comment]
I'm looking into this -- I hope to have a PR soon(ish).

Do you have any ideas on how to get the custom inspector classes into `scalac`'s classpath?

Currently, I have a custom inspector defined in `project/src/main/scala/my/inspections/MyInspection.scala`. This class is then available to the SBT build process, but it's not available by default in the `scalac` process. (The same problem would apply if the custom inspector was provided in an external JAR -- I'll make sure the docs for this feature cover both options.)

This (old) thread has some suggestions: http://www.scala-lang.org/old/node/6664 but none of them look like they'll work here. I haven't found any other documentation for this.

I'll poke about inside SBT's internals ("AggressiveCompile.scala") to see if I can figure out a way to add to the `scalac` classpath.

Thanks

**Comment by RichardBradley** [comment]
Unless you have a better idea, I'm adding a new plugin option 
```
    "-P:scapegoat:customInspectionsClasspath:<paths>      colon separated list of classpaths from which",
    "                                                     to load custom inspections",
```

**Comment by sksamuel** [comment]
I can't think of a better way either. The ideal solution would be to pass a special classpath to scalac that is used for compiler plugins, but I don't think you can do that (at least I couldn't find anything just no).

**Comment by RichardBradley** [comment]
Hi, @sksamuel :

a) Did you have a look at the PR for this in #114 and sksamuel/sbt-scapegoat#27 ?
I think they're complete enough for merging, although they're not 100% polished.
Are you happy with the design so far?
Is there anything under my control which would prevent you from merging it?

b) I have been thinking about [this issue](https://github.com/sksamuel/sbt-scapegoat/pull/27/files#diff-5a749244d1e091ce0adda24da548f08dR17) whereby the user must list all Inspections in the JARs by classname.

The best fix I can think of would be for `scalac-scapegoat` to scan all classes provided in `-P:scapegoat:customInspectionsClasspath`, and load each one, and check if each one implements `Inspection`.

This might be a surprising, expensive or risky thing for the plugin to do, but I think it's the best fix. What do you think?

I don't think it should be expensive: we're only loading a few classes; nor particularly risky as the `customInspectionsClasspath` is trusted and under user control. However, I don't think it's a "normal" thing to do in a JVM plugin. I think most JVM plugin APIs require plugin JARs to explicitly list the exported classes in a resource text file or similar.

Would you be OK with `scapegoat` scanning the user-provided JARs like this? It seems the most helpful thing to do from a UX point of view.

**Comment by RichardBradley** [comment]
>  I think most JVM plugin APIs require plugin JARs to explicitly list the exported classes in a resource text file or similar.

Examples:
 * https://code.google.com/p/findbugs/wiki/DetectorPluginTutorial#Building
 * https://en.wikipedia.org/wiki/OSGi#Bundles
 * http://docs.oracle.com/javase/6/docs/api/java/util/ServiceLoader.html
 * http://jpf.sourceforge.net/index.html

I think I can make this work. 
`scapegoat` could offer compile-time support for getting the list of inspections in the resource file correct -- an Inspection that checks that your `src/main/resources/META-INF/services/com.sksamuel.scapegoat.Inspection` file contains all instances of `Inspection` in your codebase.

**Comment by RichardBradley** [comment]
> scapegoat could offer compile-time support for getting the list of inspections in the resource file correct 

I haven't come up with anything good for this yet: http://stackoverflow.com/questions/33281999/automatic-meta-inf-services-generation-in-scala-and-sbt-for-serviceloader

**Comment by cchantep** [comment]
Anythin up there?

**Comment by RichardBradley** [comment]
This is ready and suitable for merging from my POV.
If you really want this feature, you could work from my branch while a decision is pending.

**Comment by fedeoasi** [comment]
This is pretty cool. Any idea of when https://github.com/sksamuel/scapegoat/pull/114 will be accepted and released? I see that it needs merging.

**Comment by mwz** [comment]
This sounds quite good, any chance to resurrect this @RichardBradley?

**Comment by RichardBradley** [comment]
My comment above (Dec 16, 2016) is still the latest on this as far as I know, i.e.

> This is ready and suitable for merging from my POV.
> If you really want this feature, you could work from my branch while a decision is pending \[from sksamuel\].

I haven't been using `scapegoat` recently, so I don't know if this branch still works, but it worked in 2016
